Cosine.lua

```diff
@@ -0,0 +1,163 @@
+      self._weight:addr(1, self._weight, -1/(self.__norm*self.__norm), self.output, input)
+      self.gradInput:addmv(0, 1, self._weight:t(), gradOutput)
+   elseif input:dim() == 2 then  
+      local inputNorm = self._inputNorm:expandAs(input)
+      local weightNorm = self._weightNorm:view(1,outputSize):expandAs(gradOutput)
+      
+      self.gradInput:copy(input):cdiv(inputNorm)
+      self._gradOutput = self._gradOutput or gradOutput.new()
+      self._gradOutput:resizeAs(gradOutput):copy(gradOutput)
+      self._gradOutput:cmul(self.output)
+      self._sum = self._sum or input.new()
+      self._sum:sum(self._gradOutput, 2)
+      self.gradInput:cmul(self._sum:expandAs(input))
+      
+      self._gradOutput:resizeAs(gradOutput):copy(gradOutput)
+      self._gradOutput:cdiv(weightNorm)
+      self.gradInput:addmm(-1, self.gradInput, 1, self._gradOutput, self.weight)
+      
+      self.gradInput:cdiv(inputNorm)
+   end
+
+   return self.gradInput
+end
+
+function Cosine:accGradParameters(input, gradOutput, scale)
+   scale = scale or 1
+   local inputSize = self.weight:size(2)
+   local outputSize = self.weight:size(1)
+   
+   --[[ 
+   dy_j            x_i                     w_ji
+   ----- = -------------------  -  y_j -----------
+   dw_ji   || w_j || * || x ||         || w_j ||^2
+   --]]
+   
+   if input:dim() == 1 then
+      self._gradOutput = self._gradOutput or gradOutput.new()
+      self._gradOutput:resizeAs(gradOutput):copy(gradOutput)
+      local weightNorm = self._weightNorm:view(outputSize)
+      self._gradOutput:cdiv(weightNorm)
+      self.gradWeight:addr(scale/self.__norm, self._gradOutput, input)
+      
+      self._gradOutput:cdiv(weightNorm)
+      self._gradOutput:cmul(self.output)
+      self._weight = self._weight or self.weight.new()
+      self._weight:resizeAs(self._weight):copy(self.weight)
+      self._weight:cmul(self._gradOutput:view(outputSize, 1):expandAs(self.weight))
+      self.gradWeight:add(-1, self._weight)
+   elseif input:dim() == 2 then
+      self._weight = self._weight or self.weight.new()
+      self._weight:resizeAs(self.weight):copy(self.weight)
+      self._gradOutput = self._gradOutput or gradOutput.new()
+      self._gradOutput:resizeAs(gradOutput):copy(gradOutput)
+      self._gradOutput:cmul(self.output)
+      self._sum = self._sum or input.new()
+      self._sum:sum(self._gradOutput, 1)
+      local grad = self._sum[1]
+      grad:cdiv(self._weightNorm:select(2,1))
+      self._weight:cmul(grad:view(outputSize,1):expandAs(self._weight))
+      
+      local input_ = self._gradOutput
+      input_:resizeAs(input):copy(input)
+      input_:cdiv(self._inputNorm:expandAs(input))
+      self._weight:addmm(-1, self._weight, 1, gradOutput:t(), input_)
+      
+      self._weight:cdiv(self._weightNorm:expandAs(self._weight))
+      self.gradWeight:add(self._weight)
+   else
+      error"1D or 2D input expected"
+   end
+end
+
+function Cosine:type(type)
```

Comment:
Just noticed that this module needs the `tensorCache` parameter to properly share it's weights, as per https://github.com/torch/nn/issues/415